### PR TITLE
Autolcicking on holding mouse and swap hands fix

### PR DESCRIPTION
**Pull request by Razharas**
Added preference setting that makes you able to just hold down mouse
button and if you have something in your hands it will autoclick on
everything you are pointing at, no need to clickspam anymore, take care
of your fingers and mouses

Made darkness clickable, so you cna shoot at it and swap hands while clicking at it with middle mouse

Thats it

### CODE REVIEW

**Comment by Iamgoofball** [comment]
oh shit

**Comment by MrPerson** [comment]
Does autoclick REALLY need to be a client pref? Who the fuck is going to turn it off?

**Comment by Razharas** [comment]
People who hate coders

File: code/modules/client/client procs.dm [hunk 1]
```diff
@@ -20,6 +20,34 @@
...
+/client/MouseDown(object,location,control,params)
+	if(mob && prefs && prefs.autoclick_delay && object && istype(mob, /mob/living/carbon))
+		var/mob/living/carbon/C = mob
+		if(!C.get_active_hand())
+			return ..()
+		C.is_mouse_down = object
+		spawn()
+			while(C.is_mouse_down == object)
+				C.ClickOn(C.is_mouse_down, params)
+				sleep(1)
```
Comment (MrPerson) [inline code/modules/client/client procs.dm]: This should be sleep(world.tick_lag) so it'll work correctly if we move to 20+ fps.

**Comment by Iamgoofball** [comment]
for people bitching "m-m-muh autoharm-batoning!!!"

this is the same as if I had just spammed clicks manually

all this does is save me from getting carpal tunnel syndrome

@MrPerson When I'm trying to do fine precision shit and don't want to risk accidentally clicking multiple times, yes, turning it off would be nice. But, for combat, yeah, having this is beautiful.

File: code/modules/client/client procs.dm (hunk 1 above)
Comment (Razharas) [inline code/modules/client/client procs.dm]: It actually should be AUTOCLICK_DELAY define which i made but forgot to use

File: code/modules/client/client procs.dm [hunk 2]
```diff
@@ -20,6 +20,34 @@
...
+			return
+
+/client/MouseDrag(src_object,over_object,src_location,over_location,src_control,over_control,params)
+	if(over_object)
+		MouseDown(over_object, over_location, over_control, params)
```
Comment (MrPerson) [inline code/modules/client/client procs.dm]: Is this really what we want? This'll totally mess up any other MouseDrag() actions

**Comment by MrPerson** [comment]
@Iamgoofball What kind of precision things? I feel we should make those things less shitty. There's also no in-game way to toggle the pref and even if there was, fix the damn problem rather than working around it!

File: code/modules/client/client procs.dm (hunk 2 above)
Comment (Razharas) [inline code/modules/client/client procs.dm]: I know, it only works if you have something in hands, so if your hand is free its as usual
Not much can be done about it really, no way to know if you want to pull someones icon onto yourself you beat them repeatedly

**Comment by Razharas** [comment]
Can add the pref change into prefs tab

**Comment by vista-** [comment]
If this gets in, it's gonna change robust as we know it forever.

**Comment by Aranclanos** [comment]
>client preference

why

**Comment by Razharas** [comment]
What else?
Clearly this being the only way is not good since all kinds of drag and drop iteractions

**Comment by PKPenguin321** [comment]
>Made it so you can swap hands with middle mouse button even if clicking
on the empty void of nothingness

i love you

**Comment by JJRcop** [comment]
What about if you double click and hold? Then it shows you're very intent on autolcicking, so the pref might not even be necessary, and it solves the drag n' drop conflict.

**Comment by Aranclanos** [comment]
its awful don't do this, use the obj/screen I already told how to do it

**Comment by Dorsisdwarf** [comment]
Is autoclick delay the same as regular clickdelay?

**Comment by Razharas** [comment]
You attempt to click at frame rate speed but once you did hit something usual attack delay happens

**Comment by freerealestate** [comment]
I love you

**Comment by Razharas** [comment]
It can but:
1 - i want to leave a method to quickly get rid of the autoclick by just dropping hand
2 - this will boost weapons of any kind against unarmed strongly, with i like, disarm is too good

**Comment by optimumtact** [comment]
:-1: get good

File: code/__HELPERS/unsorted.dm [hunk 3]
```diff
@@ -1304,3 +1304,14 @@ var/list/WALLITEMS = list(
...
 			return "white"
+
+/proc/screen_loc2turf(scr_loc, turf/origin)
+	var/tX = text2list(scr_loc, ",")
+	var/tY = text2list(tX[2], ":")
+	var/tZ = origin.z
+	tY = tY[1]
+	tX = text2list(tX[1], ":")
+	tX = tX[1]
+	tX = max(1, min(256, origin.x + (text2num(tX) - (world.view + 1))))
```
Comment (RemieRichards) [inline code/__HELPERS/unsorted.dm]: 256 smells like something that should be world.maxX and world.maxY.

**Comment by Aranclanos** [comment]
I love the autoclick for construction, it makes it stupidly faster for things like reflooring, there are some issues for when using the same object twice, like metal rods in a space turf, but its not a big deal.

Buuuuut at the same time it completely breaks combat, by changing how its played and almost securing a hit each time you run close to someone. This ends up with combat taking less time and without enough space to use the surroundings, which I believe makes the experience more enjoyable, but I guess that's just an opinion.


I don't know why the swap hands fix is tied to this pr, but you're using a screen obj that should be merged with the other two and make them be just overlays, making this future change harder and still using mouseup. Add a new screen obj.

**Comment by Razharas** [comment]
There are a lot of screen manipulations in the code everywhere, i tried adding new object directly in the new of client but it gets cut out somewhere and im not sure it wont be cut out again by some event
I really dont want to rewrite client screen handling

**Comment by kingofkosmos** [comment]
:+1: Yes please to autoclick in combat so the gap between "robust" <0.1 ping and non-american >0.2 ping players can be made smaller.

**Comment by Aranclanos** [comment]
no matter how much razh asks, don't merge this, autoclick on fights is shit

**Comment by Razharas** [comment]
Ahahaha good joke aran you got me

**Comment by MrStonedOne** [comment]
no no no no no no no.....


There is gonna be too many accidental gun fires when people meant to do other actions.

This is just gonna end up being a nightmare

**Comment by freerealestate** [comment]
nope

File: tgstation.dme [hunk 4]
```diff
@@ -6,6 +6,52 @@
 
 // BEGIN_FILE_DIR
 #define FILE_DIR .
+#define FILE_DIR "html"
+#define FILE_DIR "icons"
+#define FILE_DIR "icons/ass"
+#define FILE_DIR "icons/effects"
+#define FILE_DIR "icons/mecha"
```
Comment (Cheridan) [inline tgstation.dme]: wowwww

**Comment by Razharas** [comment]
You can turn it off
You can drop anything in hand

File: tgstation.dme (hunk 4 above)
Comment (Razharas) [inline tgstation.dme]: Fuck this shit

Comment (Iamgoofball) [inline tgstation.dme]: 10/10

**Comment by vista-** [comment]
@MrStonedOne, that only makes this change even better. Now security can be proper incompetent.

**Comment by optimumtact** [comment]
Please merge this to stop raz merge begging

**Comment by Razharas** [comment]
Oranges will donate 10$ to anyone who will merge this

**Comment by Aranclanos** [comment]
IM PRETTY SURE THATS CORRPUTION RAZH

**Comment by Iamgoofball** [comment]
IM PRETTY SURE THE MONEY WILL JUST HAPPEN TO BE GIVEN TO THE PERSON WHO MERGES IT COMPLETELY UNRELATED TO THE MERGE ARAN

**Comment by Razharas** [comment]
Theres literally nothing wrong with corruption
Putin told me so

**Comment by kingofkosmos** [comment]
Let's just have an eternal test period with this like any other controversial pr.

**Comment by Razharas** [comment]
^This

**Comment by Aranclanos** [comment]
ok set up a test with mso using this branch instead of merging on master

**Comment by Jordie0608** [comment]
This will probably fix #5386

**Comment by Razharas** [comment]
It will 100%

**Comment by Razharas** [comment]
Looks like its over

**Comment by Iamgoofball** [comment]
redo but with disables on combat so we engineers don't get carpal tunnel

**Comment by kingofkosmos** [comment]
what why no :grimacing:

**Comment by freerealestate** [comment]
but why?

this is needed to give people with shit latencies a fighting chance.

**Comment by PKPenguin321** [comment]
nooo

at the very least make another PR with just the middleclick on darkness fix thing

**Pull request closed without merging**